File: 2160-minimum-operations-to-make-a-uni-value-grid/minimum-operations-to-make-a-uni-value-grid.py
class Solution:
    def minOperations(self, grid: List[List[int]], x: int) -> int:
        # brute force one can easly apply change on one element and recurse on others
        # grid is not needed, it could be one array
        # intutn - minimum steps would be somewhere between min amd max, maybe no
        # could we bin search for min element ? no as # of operations is no incr
        # in fact its a curve with lowest point as answer
        # is it true that uni val is always one of the elements of array ? yes , may be no
        # if there are two elements with same is that uni value - not necessary 
        # imagine 1,2,3,5,5 - 1 - making uni val as 3 we get 7 but unival as 5 we get 9 ops
        # line sweep ? 
        # how about dp - T[] = T[]
        # of all states possible ans lies in one path to node
        # we do not know dest node or path 
        # bfs 
        # after first hint, substraction is division -> 
        # 2 + 2k1, 4 + 2k2 . . . 6 + 2k3 , 8 + 2k4
        # 2 - 2k1 , ...
        nums_array = []
        result = 0

        # Flatten the grid into nums_array
        for row in grid:
            for num in row:
                nums_array.append(num)

        # Sort nums_array in non-decreasing order to easily find the median
        nums_array.sort()

        length = len(nums_array)
        # Store the median element as the final common value
        final_common_number = nums_array[length // 2]

        # Iterate through each number in nums_array
        for number in nums_array:
            # If the remainder when divided by x is different, return -1
            if number % x != final_common_number % x:
                return -1
            # Add the number of operations required to make the current number equal to final_common_number
            result += abs(final_common_number - number) // x

        return result
        

File: 2160-minimum-operations-to-make-a-uni-value-grid/minimum-operations-to-make-a-uni-value-grid.rs
use std::cmp::min;
use std::collections::HashSet;
impl Solution {
    pub fn min_operations(grid: Vec<Vec<i32>>, x: i32) -> i32 {

        /* its always apossible ? No ,shortest path is just  step behind
        many destinations ?
        maybe mmost frequent element is an unique number
        one of the number from array is unique number
        grid is not relevant ? it could be an array ?
        could binary search be possible ?
        we search for ans ?
        sum(arr) + knx // m is an integer ?

        m0st promising is choose each number from gid
        then find the diff between then
        [2,4,6,8] -> [4,4,4,4] 16-12 = 4 which means 2x
        [2,0,2,2*2] // not this does not work as plus 2 and -2 are eaten by each other and we never find ans
        given that 10000 is the limit of 
        choose 1 to 10K
        for each select and element as answer and perform the op on array ?
        */
        let mut hs = HashSet::new();
        let mut a = 0;
        let mut new_arr = Vec::new();

        
         for arr in grid.clone(){
                for j in arr {
                    new_arr.push(j);
                    a = j;
                    hs.insert(a);
                }
         }

         if new_arr.len() == 1 {
            return 0;
         }

         new_arr.sort();
         let dest = new_arr[new_arr.len() / 2];
         
        let mut res = 10000000;
        let mut temp = 0;
        let mut invalid = false;
        for j in new_arr.clone(){
                        //print!("{} {} {}\n", k as i32,j,x);
            if ((dest - j) % x != 0) {
                temp = 10000000;
                break;
            }
            temp += ((dest - j) / x).abs();
                       
        }
        res = temp;
                //print!("{}\n", temp as i32);
        if res == 10000000 {
            return -1;
        }
        return res;
        
    }
}